### PR TITLE
Supabase hybrid keyword search

### PR DESCRIPTION
**Pull request by ericciarla**

### CODE REVIEW

File: langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  tableName: string;
+
+  kwQueryName: string;
+
+  embeddings: Embeddings;
+
+  constructor(
+    embeddings: Embeddings,
+    args: SupabaseLibArgs,
+    sim_k: number,
```
Comment (nfcampos) [inline langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts:63]: Let's move `sim_k` and `kw_k` into `args`

File: langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    const result: [Document, number][] = (
+      searches as SearchEmbeddingsResponse[]
+    ).map((resp) => [
+      new Document({
+        metadata: resp.metadata,
+        pageContent: resp.content,
+      }),
+      resp.similarity,
+    ]);
+    console.log(result);
```
Comment (nfcampos) [inline langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts:103]: Remove

File: langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    const result: [Document, number][] = (searches as kwSearchResponse[]).map(
+      (resp) => [
+        new Document({
+          metadata: resp.metadata,
+          pageContent: resp.content,
+        }),
+        resp.similarity * 10,
+      ]
+    );
+    console.log(result);
```
Comment (nfcampos) [inline langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts:131]: Remove

File: langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    return result;
+  }
+
+  async SupabaseHybridKeyWordSearch(
```
Comment (nfcampos) [inline langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts:135]: A method name starting with a capital letter looks a little odd, I think we should rename

File: langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+      throw new Error(`Error searching for documents: ${error}`);
+    }
+
+    const result: [Document, number][] = (searches as kwSearchResponse[]).map(
+      (resp) => [
+        new Document({
+          metadata: resp.metadata,
+          pageContent: resp.content,
+        }),
+        resp.similarity * 10,
```
Comment (nfcampos) [inline langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts:128]: Why `*10`?

File: langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    kw_k: number
+  ): Promise<[Document, number][]> {
+    const simularity_search = this.similaritySearchVectorWithScore(
+      await this.embeddings.embedQuery(query),
+      sim_k
+    );
+    const keyword_search = this.keywordSearch(query, kw_k);
+
+    return Promise.all<[Document, number][]>([
+      simularity_search,
```
Comment (nfcampos) [inline langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts:147]: typo

File: langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    return result;
+  }
+
+  async SupabaseHybridKeyWordSearch(
+    query: string,
+    sim_k: number,
+    kw_k: number
+  ): Promise<[Document, number][]> {
+    const simularity_search = this.similaritySearchVectorWithScore(
+      await this.embeddings.embedQuery(query),
```
Comment (nfcampos) [inline langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts:141]: I'd say we can move this inside `similaritySearchVectorWithScore` and rename it to `similaritySearch` which would make it more consistent with `keywordSearch`

File: langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts (hunk 4 above)
Comment (nfcampos) [reply langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts:135]: Maybe rename to just `hybridSearch`

**Review by nfcampos — COMMENTED** [review]
A few comments, looks good otherwise, thanks!

File: langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts (hunk 4 above)
Comment (ericciarla) [reply langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts:133]: Okay I agree, that looks good!

File: langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts (hunk 7 above)
Comment (ericciarla) [reply langchain/src/retrievers/SupabaseHybridKeyWordSearch.ts:139]: Okay all set!

File: langchain/src/retrievers/supabase-hybrid-search.ts [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+    return Promise.all<[Document, number][]>([
+      similarity_search,
+      keyword_search,
+    ])
+      .then((results) => results.flat())
+      .then((results) => {
+        const seenContent = new Set();
+        return results.filter(([doc]) => {
+          const content = doc.pageContent;
```
Comment (nfcampos) [inline langchain/src/retrievers/supabase-hybrid-search.ts:158]: @ericciarla one last comment, I think this should either take the max or the sum of the scores for documents found in both searches, thoughts?

Comment (ericciarla) [reply langchain/src/retrievers/supabase-hybrid-search.ts:158]: @nfcampos I am alright with that. I think max score would be best. Is that some you want to quickly add? I can as well, just let me know

File: docs/docs/modules/indexes/retrievers/supabase_hybrid.mdx [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    content,
+    metadata,
+    1 - (documents.embedding <=> query_embedding) as similarity
+  from documents
+  order by documents.embedding <=> query_embedding
+  limit match_count;
+end;
+$$;
+
+-- Create an index to be used by the similarity search function
```
Comment (mmmaia) [inline docs/docs/modules/indexes/retrievers/supabase_hybrid.mdx:55]: @ericciarla Indexes should be created after the table has data.
https://github.com/pgvector/pgvector#indexing

**Pull request merged**